src/core/egg-herbie.rkt

```diff
@@ -125,39 +124,86 @@
       [(? symbol? x) (egraph_add_node ptr (symbol->string x) 0-vec root?)]
       [(? number? n) (egraph_add_node ptr (number->string n) 0-vec root?)]))
 
+  ; The function recurses on spec
+  (define (batch-parse-approx batch)
+    (batch-replace batch
+                   (lambda (node)
+                     (match node
+                       [(approx spec impl) (list '$approx spec impl)]
+                       [_ node]))))
+
+  (set-batch-roots! batch roots) ; make sure that we work with the right roots
+  ; the algorithm may crash if batch-length is zero
+  (define insert-batch
+    (if (zero? (batch-length batch)) batch (remove-zombie-nodes (batch-parse-approx batch))))
+
+  (define mappings (build-vector (batch-length insert-batch) values))
+  (define (remap x)
+    (vector-ref mappings x))
+
+  ; Inserting nodes bottom-up
+  (define root-mask (make-vector (batch-length insert-batch) #f))
+  (for ([root (in-vector (batch-roots insert-batch))])
+    (vector-set! root-mask root #t))
+  (for ([node (in-vector (batch-nodes insert-batch))]
+        [root? (in-vector root-mask)]
+        [n (in-naturals)])
+    (define node*
+      (match node
+        [(literal v _) v]
+        [(? number?) node]
+        [(? symbol?) (normalize-var node)]
+        [(list '$approx spec impl)
+         (hash-ref! id->spec
+                    (remap spec)
+                    (lambda ()
+                      (define spec* (normalize-spec (batch-ref insert-batch spec)))
+                      (define type (representation-type (repr-of-node insert-batch impl ctx)))
+                      (cons spec* type))) ; preserved spec and type for extraction
+         (list '$approx (remap spec) (remap impl))]
+        [(list op (app remap args) ...) (cons op args)]))
+
+    (vector-set! mappings n (insert-node! node* root?)))
+
+  ;------------------------- DEBUGGING
```

Comment:
This debugging code does the same functionality for adding nodes to an `egg`, but it is recursion-based. It is here for checking whether accuracy indeed can change by switching to recursion